src/schema.ts

```diff
@@ -173,7 +181,281 @@ class SchemaDefinition implements SchemaDefinitionType {
+        return Object.values(this.getProperties(modelName))
+            .filter((propDefn) => !propDefn.mandatory)
+            .map((propDefn) => propDefn.name);
+    }
+
+    getProperty(modelName: string, property: string): PropertyDefinition {
+        return this.getProperties(modelName)[property];
+    }
+
+    hasProperty(modelName: string, property: string): boolean {
+        return Boolean(this.getProperty(modelName, property));
+    }
+
+    inheritsFrom(modelName: string, parent: string): boolean {
+        return this.ancestors(modelName).includes(parent);
+    }
+
+    /**
+     * Returns the list of properties for the Class.active index. This is
+     * expected to represent a unique constraint on a combination of these properties
+     *
+     * @returns list of properties which belong to the index
+     */
+    activeProperties(modelName: string): string[] | null {
+        const model = this.get(modelName);
+
+        for (const index of model.indices) {
+            if (index.name === `${model.name}.active`) {
+                return index.properties;
+            }
+        }
+        return null;
+    }
+
+    /**
+     * the list of parent class names which this class inherits from
+     */
+    ancestors(modelName: string): string[] {
+        const model = this.get(modelName);
+        const parents: string[] = [];
+
+        for (const parent of model.inherits) {
+            parents.push(parent);
+            parents.push(...this.ancestors(parent));
+        }
+        return parents;
+    }
+
+    children(modelName: string): string[] {
+        const model = this.get(modelName);
+        return this.subclassMapping[model.name] || [];
+    }
+
+    /**
+     * Get a list of models (including the current model) for all descendants of the current model
+     *
+     * @param excludeAbstract exclude abstract models
+     *
+     * @returns the array of descendant models
+     */
+    descendants(
+        modelName: string,
+        opt: { excludeAbstract?: boolean; includeSelf?: boolean } = {},
+    ): string[] {
+        const { excludeAbstract = false, includeSelf = false } = opt;
+        const descendants: string[] = includeSelf
+            ? [modelName]
+            : [];
+        const queue: string[] = this.children(modelName).slice();
+
+        while (queue.length > 0) {
+            const child = queue.shift() as typeof queue[number];
+
+            if (descendants.includes(child)) {
+                continue;
+            }
+            descendants.push(child);
+            queue.push(...this.children(child));
+        }
+        return descendants.filter((child) => {
+            const descendantModel = this.get(child, true);
+            return (!excludeAbstract || !descendantModel.isAbstract);
+        });
+    }
+
+    /**
+     * Returns a set of properties from this class and all subclasses
+     * @type {Array.<PropertyDefinition>}
+     */
+    queryableProperties(modelName: string) {
```

Comment:
missing return type

Reply:
I had been leaving them off when they could be inferred based on the types already there. Should I be adding them explicitly anyway? vscode/ts adds the return type automatically b/c we have it declared for this.getProperties already

Reply:
I guess it's not strictly necessary. I like having them for functions as it makes it clear what you expect to be returned is what's being returned and if later some logic within the function was changed that affected the return type it would be clear